### PR TITLE
History fix cursor update

### PR DESCRIPTION
**Pull request by sherubthakur**
Various fixes to history execution

### CODE REVIEW

**Comment by sholderbach** [comment]
Still some usability painpoints:

* Ctrl-R search has the refresh bugs
* moving the cursor left/right without editing breaks the ability to go up/down
* still skips the most recent entry going up
* you can not return to the (empty) linebuffer by going down beyond the history (some shells like bash allow you to return to a partial manual entry if they don't support the prefix searching history browsing)

For the last two points the iterator like interface of moving while either returning Some(entry) when a move was succesful or None when exhausted had some benefits over the new setup. We want to be able to handle those cases. I started some changes on that, if I get them to work I will file a PR.

**Comment by sholderbach** [comment]
OK Update regarding point two on my list. zsh's history traversal would also stop at this point as it then thinks it is in prefix mode. Yet bash allows horizontal cursor movement without interfering with the basic up/down history navigation. So this seems to be somewhat of an UI/UX choice.

**Comment by sherubthakur** [comment]
> Ctrl-R search has the refresh bugs

IIRC this has existed for a long time, I wasn't really focused on this one. I really think we should get rid of the clock update for now and introduce it later down the road.

> moving the cursor left/right without editing breaks the ability to go up/down

Can you elaborate on this one (ideally with an example). So, I can understand what the behaviour should be. I believe the present state (in the PR) should be that left/right movement etc should stop history searching altogether for Normal and PrefixSearch modes. That is what I wanted to implement.
Though I think when doing `crtl-r` search this should be allowed (i.e. going left/right). (Not sure what the present behaviour is) 

> still skips the most recent entry going up

Yeah, I plan to fix this one.

> ou can not return to the (empty) line buffer by going down beyond the history (some shells like bash allow you to return to a partial manual entry if they don't support the prefix searching history browsing)

I had not really thought this through. Definitely, something that should happen.

**Comment by jntrnr** [comment]
@sherubthakur - taking a look at this now. I'm seeing something related to a previous PR you did to clean up history and I think it'd be great to fix it as part of this update.

If you press up/down and wait a second, the input disappears. Before #92 this worked fine, but after that it seems that after the second ticks by we only draw the prompt and not the user's input.

File: src/line_buffer.rs [hunk 1]
```diff
@@ -65,8 +65,17 @@ impl LineBuffer {
 
     /// Set to a single line of `buffer` and reset the `InsertionPoint` cursor
     pub fn set_buffer(&mut self, buffer: String) {
```
Comment (sholderbach) [inline src/line_buffer.rs:67]: Are other users of `set_buffer` maybe expecting a different behavior? Else we may want to have have `set_buffer_cursor_end` or something else.

File: src/engine.rs [hunk 2]
```diff
@@ -337,7 +337,13 @@ impl Reedline {
             (true, _) => {
                 self.history.set_navigation(HistoryNavigationQuery::Normal);
             }
-            (false, HistoryNavigationQuery::PrefixSearch(_)) => {}
+            (false, HistoryNavigationQuery::PrefixSearch(prefix)) => {
```
Comment (sholderbach) [inline src/engine.rs:340]: Out of curiosity: Against which state is this guarding?

**Review by sholderbach — COMMENTED** [review]
Let's get this merged that we can remove this history bug, and check for the new interactions.

If the refactored line buffer will have to be taken in to account depends on what gets merged first.

**Comment by jntrnr** [comment]
Yeah I'm also :+1: on landing and continuing to polish. @sherubthakur thanks for the fixes!

File: src/engine.rs (hunk 2 above)
Comment (sherubthakur) [reply src/engine.rs:340]: Each up/down press will call this function. if the prefix does not change we don't want to reset the history state.

**Pull request merged**